File: api/test/api/responders_test/test_data/clique_constraints.py
###############################################################################
# Copyright (c) 2017-2019 Koren Lev (Cisco Systems),                          #
# Yaron Yogev (Cisco Systems), Ilia Abashin (Cisco Systems) and others        #
#                                                                             #
# All rights reserved. This program and the accompanying materials            #
# are made available under the terms of the Apache License, Version 2.0       #
# which accompanies this distribution, and is available at                    #
# http://www.apache.org/licenses/LICENSE-2.0                                  #
###############################################################################
from api.test.api.responders_test.test_data import base

URL = "/clique_constraints"

WRONG_ID = base.WRONG_OBJECT_ID
NONEXISTENT_ID = "576a4176a83d5313f21971f0"
CORRECT_ID = base.CORRECT_OBJECT_ID

WRONG_FOCAL_POINT_TYPE = base.WRONG_OBJECT_TYPE
CORRECT_FOCAL_POINT_TYPE = base.CORRECT_OBJECT_TYPE

CONSTRAINT = "network"

CLIQUE_CONSTRAINTS_WITH_SPECIFIC_ID = [
    {
       "id": CORRECT_ID
    }
]

CLIQUE_CONSTRAINTS_WITH_SPECIFIC_FOCAL_POINT_TYPE = [
    {
       "id": "576a4176a83d5313f21971f5",
       "focal_point_type": CORRECT_FOCAL_POINT_TYPE
    },
    {
        "id": "576ac7069f6ba3074882b2eb",
        "focal_point_type": CORRECT_FOCAL_POINT_TYPE
    }
]

CLIQUE_CONSTRAINTS_WITH_SPECIFIC_FOCAL_POINT_TYPE_RESPONSE = {
    "clique_constraints": CLIQUE_CONSTRAINTS_WITH_SPECIFIC_FOCAL_POINT_TYPE
}

CLIQUE_CONSTRAINTS_WITH_SPECIFIC_CONSTRAINT = [
    {
        "id": "576a4176a83d5313f21971f5",
        "constraints": [
            CONSTRAINT
        ]
    },
    {
        "id": "576ac7069f6ba3074882b2eb",
        "constraints": [
            CONSTRAINT
        ]
    }
]

CLIQUE_CONSTRAINTS_WITH_SPECIFIC_CONSTRAINT_RESPONSE = {
    "clique_constraints": CLIQUE_CONSTRAINTS_WITH_SPECIFIC_CONSTRAINT
}

CLIQUE_CONSTRAINTS = [
    {
       "id": "576a4176a83d5313f21971f5"
    },
    {
        "id": "576ac7069f6ba3074882b2eb"
    }
]

CLIQUE_CONSTRAINTS_RESPONSE = {
    "clique_constraints": CLIQUE_CONSTRAINTS
}

File: front-end/imports/api/clique-constraints/clique-constraints.js
///////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2017-2018 Koren Lev (Cisco Systems),                         /
// Yaron Yogev (Cisco Systems), Ilia Abashin (Cisco Systems) and others       /
//                                                                            /
// All rights reserved. This program and the accompanying materials           /
// are made available under the terms of the Apache License, Version 2.0      /
// which accompanies this distribution, and is available at                   /
// http://www.apache.org/licenses/LICENSE-2.0                                 /
///////////////////////////////////////////////////////////////////////////////
import { Mongo } from 'meteor/mongo';
import { SimpleSchema } from 'meteor/aldeed:simple-schema';
import * as R from 'ramda';
import { Constants } from '/imports/api/constants/constants';
import { Environments } from '/imports/api/environments/environments';
import {callApiValidators} from "../../lib/utilities";

export const CliqueConstraints = new Mongo.Collection(
  'clique_constraints', { idGeneration: 'MONGO' });

let schema = {
  _id: { type: { _str: { type: String, regEx: SimpleSchema.RegEx.Id } } },

  focal_point_type: {
    type: String,
    custom: function () {
      let that = this;
      let values = Constants.findOne({ name: 'object_types_for_links' }).data;

      if (R.isNil(R.find(R.propEq('value', that.value), values))) {
        return 'notAllowed';
      }
    }
  },

  environment: {
    type: String,
    defaultValue: "",
    optional: true,
    custom: function () {
      let that = this;
      if (that.value !== that.definition.defaultValue) {
          let env = Environments.findOne({name: that.value});

          if (R.isNil(env)) {
              return 'notAllowed';
          }
      }
    },
  },

  constraints: {
    type: [String],
    minCount: 1,
    custom: function () {
      let that = this;
      let objectTypes = Constants.findOne({ name: 'object_types_for_links' }).data;

      let findResult = R.intersection(that.value, R.pluck('value', objectTypes));
      if (findResult.length !== that.value.length) { return 'notAllowed'; }

      // Document validators workaround
      return callApiValidators(that, [duplicateValidator]);
    },
  },
};

CliqueConstraints.schema = new SimpleSchema(schema);
CliqueConstraints.attachSchema(CliqueConstraints.schema);

function duplicateValidator(that) {
    // Validate focal point uniqueness
    console.log("Validator: duplicate constraint");
    let env = (that.field('environment').value) ? that.field('environment').value : null;

    let existing = CliqueConstraints.findOne({
        environment: env,
        focal_point_type: that.field('focal_point_type').value
    });

    if (R.allPass([
            R.pipe(R.isNil, R.not),
            R.pipe(R.propEq('_id', that.docId), R.not)
        ])(existing)) {
        console.warn("Duplicate constraint");
        return 'alreadyExists';
    }
}